modules/internal/ChapelIteratorSupport.chpl

```diff
@@ -600,7 +689,10 @@ module ChapelIteratorSupport {
 
   pragma "fn returns iterator"
   inline proc _toFastFollower(x, leaderIndex) {
+    // ENGIN: want to call chpl__canHaveFastFollowers, but we need the promotion
+    // wrapper written by the compiler.
     if chpl__staticFastFollowCheck(x) then
```

Comment:
I could be getting this wrong, but doesn't chpl__staticFastFollowCheck end up calling chpl__canHaveFastFollowers anyways?

Reply:
This was one of the main confusions for me as well and was the reason I wasn't able to pull this off in my first attempt.

`A+B+C` becomes something like `A+chpl_promo_plus` where the second operand supports fast followers, so we call `_toFastFollower` with `chpl_promo_plus`. There if we call `chpl__canHaveFastFollowers`, we need that function to be able to separate iterators and promotion expressions, and argue about the items in the promoted expression. Or, we need to write wrapper for that in the compiler. Here, I took the easy way out. If the argument is a promoted expression, then we call the compiler-written version of it that expands the promoted expression to a tuple before passing it to `chpl__staticFastFollowCheck`.

Reply:
Ah, I see. So I'm not sure if this will be directly related then, but the compiler builds up the static and dynamic fast follower checks for promoted expressions. I wonder if the compiler needs to also build up the canHaveFastFollowers check there too. I think so, and now that I'm reading it again, I think that's what your comment is saying. 

https://github.com/chapel-lang/chapel/pull/4558 added that support. I think the implementation was changed since I added it, but it should be pretty similar to that still.

Reply:
Right.

However, I can't tell whether you want to do that now. I personally think that this workaround is not too bad (I can add more to the comment there) and adding that wrapper via the compiler can be part of a bigger fast follower overhaul effort. (Reduce reliance on the module support, do more in the compiler etc)

I intended to do that but the performance benefits were a bit underwhelming so I didn't want to go that far and limit the changes to the module.

Reply:
I see. Ok, let's create a private issue for creating the `canHaveFastFollowers` in the compiler too and link to that here and maybe beef up the comment here. 

It wasn't clear to me at first, but I think it's the case that the other uses of canHaveFastFollowers are all restricted to implementations of `staticFastFollowCheck` and `dynamicFastFollowCheck` for non-promotion types so they don't screw anything up with promotion. And this check that you want to replace is the only case where we'd end up calling  `canHaveFastFollowers` on a promoted type, so that's why we can't use it here yet. 

FWIW I do not think it'd take long to implement canHaveFastFollowers in the compiler. The existing code the compiler builds up for say `staticFastFollowCheck` just forwards to the individual static checks for arrays/whatever and I think most of that infrastructure can be reused just replacing the function name.

Reply:
Eh, I had a couple minutes and wanted to see if I remembered this code. I think the following will result in building up the canHaveFastFollower checks for promoted types:

```diff
diff --git a/compiler/resolution/wrappers.cpp b/compiler/resolution/wrappers.cpp
index 21f7eec61f..75bb02d62a 100644
--- a/compiler/resolution/wrappers.cpp
+++ b/compiler/resolution/wrappers.cpp
@@ -2640,7 +2640,14 @@ static void fixUnresolvedSymExprsForPromotionWrapper(FnSymbol* wrapper,
 // exist yet), we use a primitive as a placeholder. When the record is filled
 // in during iterator lowering, we replace the primitive with the actual field.
 //
-static void buildFastFollowerCheck(bool                  isStatic,
+
+enum FastFollowerCheckType {
+  STATIC_CHECK,
+  DYNAMIC_CHECK,
+  CAN_HAVE_CHECK
+};
+
+static void buildFastFollowerCheck(FastFollowerCheckType checkType,
                                    bool                  addLead,
                                    FnSymbol*             wrapper,
                                    Type*                 IRtype,
@@ -2659,17 +2666,21 @@ static void buildFastFollowerCheck(bool                  isStatic,
   returnTmp->addFlag(FLAG_EXPR_TEMP);
   returnTmp->addFlag(FLAG_MAYBE_PARAM);

-  if (isStatic == true) {
+  if (checkType == STATIC_CHECK) {
     fnName          = "chpl__staticFastFollowCheck";

     checkFn         = new FnSymbol(fnName);
     checkFn->retTag = RET_PARAM;
-
-  } else {
+  } else if (checkType == DYNAMIC_CHECK) {
     fnName          = "chpl__dynamicFastFollowCheck";

     checkFn         = new FnSymbol(fnName);
     checkFn->retTag = RET_VALUE;
+  } else if (checkType == CAN_HAVE_CHECK) {
+    fnName          = "chpl__canHaveFastFollowers";
+
+    checkFn         = new FnSymbol(fnName);
+    checkFn->retTag = RET_PARAM;
   }

   checkFn->addFlag(FLAG_COMPILER_GENERATED);
@@ -2737,12 +2748,15 @@ void buildFastFollowerChecksIfNeeded(CallExpr* checkCall) {
   SET_LINENO(wrapFn);

   // Build static (param) fast follower check functions
-  buildFastFollowerCheck(true,  false, wrapFn, ir, requiresPromotion);
-  buildFastFollowerCheck(true,  true,  wrapFn, ir, requiresPromotion);
+  buildFastFollowerCheck(STATIC_CHECK,  false, wrapFn, ir, requiresPromotion);
+  buildFastFollowerCheck(STATIC_CHECK,  true,  wrapFn, ir, requiresPromotion);
+
+  buildFastFollowerCheck(CAN_HAVE_CHECK,  false, wrapFn, ir, requiresPromotion);
+  buildFastFollowerCheck(CAN_HAVE_CHECK,  true,  wrapFn, ir, requiresPromotion);

   // Build dynamic fast follower check functions
-  buildFastFollowerCheck(false, false, wrapFn, ir, requiresPromotion);
-  buildFastFollowerCheck(false, true,  wrapFn, ir, requiresPromotion);
+  buildFastFollowerCheck(DYNAMIC_CHECK, false, wrapFn, ir, requiresPromotion);
+  buildFastFollowerCheck(DYNAMIC_CHECK, true,  wrapFn, ir, requiresPromotion);

   // Done with this wrapFn.
   promotionFormalsMap.erase(wrapFn);
```

If that works it can probably use better enum names, and some comments need updating, but I think that general approach is what's needed to build them up.

Reply:
Cool! I'll add that and ping you again once I have the implementation.